cmd/image-autobumper/imagebumper/imagebumper.go

```diff
@@ -0,0 +1,281 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package imagebumper
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	imageRegexp = regexp.MustCompile(`\b((?:[a-z0-9]+\.)?gcr\.io|(?:[a-z0-9-]+)?docker\.pkg\.dev)/([a-z][a-z0-9-]{5,29}/[a-zA-Z0-9][a-zA-Z0-9_./-]+):([a-zA-Z0-9_.-]+)\b`)
+	tagRegexp   = regexp.MustCompile(`(v?\d{8}-(?:v\d(?:[.-]\d+)*-g)?[0-9a-f]{6,10}|latest)(-.+)?`)
+)
+
+const (
+	imageHostPart  = 1
+	imageImagePart = 2
+	imageTagPart   = 3
+	tagVersionPart = 1
+	tagExtraPart   = 2
+)
+
+type Client struct {
+	// Keys are <imageHost>/<imageName>:<currentTag>. Values are corresponding tags.
+	tagCache   map[string]string
+	httpClient *http.Client
+}
+
+func NewClient(httpClient *http.Client) *Client {
+	// Shallow copy to adjust Timeout
+	httpClientCopy := *httpClient
+	httpClientCopy.Timeout = 1 * time.Minute
+
+	return &Client{
+		tagCache:   map[string]string{},
+		httpClient: &httpClientCopy,
+	}
+}
+
+type manifest map[string]struct {
+	TimeCreatedMs string   `json:"timeCreatedMs"`
+	Tags          []string `json:"tag"`
+}
+
+// commit | tag-n-gcommit
+var commitRegexp = regexp.MustCompile(`^g?([\da-f]+)|(.+?)??(?:-(\d+)-g([\da-f]+))?$`)
+
+// DeconstructCommit separates a git describe commit into its parts.
+
+// Examples:
+//
+//	v0.0.30-14-gdeadbeef => (v0.0.30 14 deadbeef)
+//	v0.0.30 => (v0.0.30 0 "")
+//	deadbeef => ("", 0, deadbeef)
+//
+// See man git describe.
+func DeconstructCommit(commit string) (string, int, string) {
```

Comment:
Would not it be easier to split it by `-`, then string g from the second part and do all parsing.
If not easier to do then at least easier to comprehend.

Reply:
Reworked